kong/tools/queue.lua

```diff
@@ -0,0 +1,408 @@
+--
+-- `handler` is the function which consumes entries.
+--
+-- Entries are dequeued in batches.  The maximum size of each batch of entries
+-- that is passed to the `handler` function can be configured using the
+-- `batch_max_size` parameter.
+--
+-- The handler function can either return true if the entries
+-- were successfully processed, or false and an error message to indicate that
+-- processing has failed.  If processing has failed, the queue library will
+-- automatically retry.
+--
+-- If the `batch_max_size` parameter is larger than 1, processing of the
+-- queue will only start after `max_delay` milliseconds have elapsed if
+-- less than `batch_max_size` entries are waiting on the queue.  If `batch_max_size`
+-- is 1, the queue will be processed immediately.
+--
+-- Usage:
+--
+--   local Queue = require "kong.tools.queue"
+--
+--   local handler = function(conf, entries)
+--     -- must return true if ok, or false + error otherwise
+--     return true
+--   end
+--
+--   local handler_conf = {...}  -- configuration for queue handler
+--   local queue_conf =          -- configuration for the queue itself (defaults shown unless noted)
+--     {
+--       name = "example",       -- name of the queue (required)
+--       batch_max_size = 10,    -- maximum number of entries in one batch (default 1)
+--       max_delay = 1,          -- maximum number of seconds after first entry before a batch is sent
+--       capacity = 10,          -- maximum number of entries on the queue (default 10000)
+--       string_capacity = 100,  -- maximum number of bytes on the queue (default nil)
+--       max_retry_time = 60,    -- maximum number of seconds before a failed batch is dropped
+--       max_retry_delay = 60,   -- maximum delay between send attempts
+--     }
+--
+--   Queue.enqueue(queue_conf, handler, handler_conf, "Some value")
+--   Queue.enqueue(queue_conf, handler, handler_conf, "Another value")
+--
+-- Given the example above,
+--
+-- * If the two `enqueue()` invocations are done within `max_delay` seconds, they will be passed to the
+--   handler function together in one batch.  The maximum number of entries in one batch is defined
+--   by the `batch_max_size` parameter.
+-- * The `capacity` parameter defines how many entries can be waiting on the queue for transmission.  If
+--   it is exceeded, the oldest entries on the queue will be discarded when new entries are queued.  Error
+--   messages describing how many entries were lost will be logged.
+-- * The `string_capacity` parameter, if set, indicates that no more than that number of bytes can be
+--   waiting on a queue for transmission.  If it is set, only string entries must be queued and an error
+--   message will be logged if an attempt is made to queue a non-string entry.
+-- * When the `handler` function does not return a true value for a batch, it is retried for up to
+--   `max_retry_time` seconds before the batch is deleted and an error is logged.  Retries are organized
```

Comment:
Is the batch deleted even if the queue is not full or near capacity?

Reply:
That is a bit of a surprise.
So, Kong will churn through all the queued items if the upstream is dead for let's say 5 minutes even if there are no requests that are sent to Kong.
Or at least a few batches (assuming exponential back off is 1 min).

Reply:
Unlimited retries when the log server is unavailable will waste CPU resources and even cause performance issues.

So we think we should consider the following trade-off.

Are we willing to trade a potential performance impact for the benefit of keeping a portion of the information when the log server is unavailable?

Reply:
Limiting the number of retries was in the original queue design and has been carried over.  I am not sure if there was a lot of discussion around the design choice at the time.  With the new `max_retry_seconds` parameter, it is possible to directly influence how long Kong retries, so if users want to be able to keep old entries, they can set the parameter to a very long timeout.  We could also change the default to be `0` and make `0` mean that no timeout is provided, but that'd be a deviation from the original design.

One reason why users may want to drop entries that can't be sent after retrying for some time is that the problem may actually be caused by the data that is sent and that dropping would be preferable to blocking the queue forever.